### PR TITLE
Create bookDaoImpl

### PR DESCRIPTION
**Pull request by ConfusedSatlan**

### CODE REVIEW

File: src/main/resources/init_db.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE books (
+    id INT AUTO_INCREMENT PRIMARY KEY,
+    title VARCHAR(50) NOT NULL,
+    price DECIMAL(10, 2) NOT NULL
+);
```
Comment (OleksandrSaven) [inline src/main/resources/init_db.sql:5]: Add empty line

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.List;
+import java.util.Optional;
+
+@mate.academy.lib.Dao
+public class BookDaoImpl implements BookDao {
+    private static final String TABLE_NAME = "books";
+    private static final String TITLE_NAME = "title";
+    private static final String PRICE_NAME = "price";
+    private static final int FIRST_INDEX = 1;
+    private static final int SECOND_INDEX = 2;
```
Comment (OleksandrSaven) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:28]: the name of the variable should be informative

**Review by OleksandrSaven — APPROVED** [review]

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    @Override
+    public Optional<Book> findById(Long id) {
+        String sql = "SELECT * FROM " + TABLE_NAME + " WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(sql)) {
+            validateTableConnection(connection, TABLE_NAME);
+            statement.setLong(FIRST_INDEX_SQL_DATA, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                String title = resultSet.getString(TITLE_NAME);
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:64]: Do not duplicate the code. Extract mapping logic in separate private method which accepts ResultSet and use it in `finaAll()` too.

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    public Book update(Book book) {
+        String sql = "UPDATE " + TABLE_NAME + " SET title = ?, price = ? WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(sql)) {
+            validateTableConnection(connection, TABLE_NAME);
+            statement.setString(FIRST_INDEX_SQL_DATA, book.getTitle());
+            statement.setBigDecimal(SECOND_INDEX_SQL_DATA, book.getPrice());
+            statement.setLong(THIRD_INDEX_SQL_DATA, book.getId());
+            int affectedRows = statement.executeUpdate();
+            if (affectedRows < ROW_EXIST_NUM) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:111]: It's okay if zero rows are affected, do not throw an Exception.

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Override
+    public boolean deleteById(Long id) {
+        String sql = "DELETE FROM " + TABLE_NAME + " WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(sql)) {
+            validateTableConnection(connection, TABLE_NAME);
+            statement.setLong(FIRST_INDEX_SQL_DATA, id);
+            int affectedRows = statement.executeUpdate();
+            if (affectedRows < ROW_EXIST_NUM) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:128]: It's okay if zero rows are affected, do not throw an Exception. You return boolean value indicating whether 0 or more rows were affected.

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                throw new DataProcessingException("Expected to insert at leas one row, but inserted 0 row.");
+            }
+            return affectedRows >= 1;
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:131]: More correct to do like this.
```suggestion
            return affectedRows > 0;

```

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    @Override
+    public Book create(Book book) {
+        String sql = "INSERT INTO " + TABLE_NAME + " (title, price) VALUES (?, ?)";
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
+            validateTableConnection(connection, TABLE_NAME);
+            statement.setString(FIRST_INDEX_SQL_DATA, book.getTitle());
+            statement.setBigDecimal(SECOND_INDEX_SQL_DATA, book.getPrice());
+            int affectedRows = statement.executeUpdate();
+            if (affectedRows < ROW_EXIST_NUM) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:41]: It's okay if zero rows are affected, do not throw an Exception.

**Review by vovasalyha — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+             Statement statement = connection.createStatement()) {
+            ResultSet resultSet = statement.executeQuery(sql);
+            List<Book> allBooks = new ArrayList<>();
+            while (resultSet.next()) {
+                Book book = createBook(resultSet);
+                allBooks.add(book);
+            }
+            return allBooks;
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't create a connection to the DB" ,e);
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:76]: Please use `DataProcessingException` everywhere. And pay attention to error messages. Connection problems is not the only reason why exception can be thrown, think about better message for this context.

File: src/main/resources/init_db.sql [hunk 9]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE books (
+    id INT AUTO_INCREMENT PRIMARY KEY,
+    title VARCHAR(50) NOT NULL,
```
Comment (vovasalyha) [inline src/main/resources/init_db.sql:3]: It's too restrictive length IMO. There are books with longer names for sure.

File: src/main/java/mate/academy/dao/BookDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        return Optional.empty();
+    }
+
+    @Override
+    public List<Book> findAll() {
+        String sql = "SELECT * FROM " + TABLE_NAME;
+        try (Connection connection = ConnectionUtil.getConnection();
+             Statement statement = connection.createStatement()) {
+            ResultSet resultSet = statement.executeQuery(sql);
+            List<Book> allBooks = new ArrayList<>();
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dao/BookDaoImpl.java:69]:
```suggestion
            List<Book> books = new ArrayList<>();

```

**Review by vovasalyha — APPROVED** [review]

**Pull request closed without merging**